### ISSUE TITLE
Switch to ReadTheDocs

### ISSUE THREAD

**Issue by wouterbeek** [TriplyDB/Documentation#74]
# Requirements

- [x] The currently visited page is not highlighted in the left-hand menu.

![Screenshot from 2023-07-21 08-19-56](https://github.com/TriplyDB/Documentation/assets/1685012/9b7cc8e5-93c5-493b-ae72-8f335a049ace)

- [x] The chapters inside a page (h2) are not visible in the left-hand menu. This makes it difficult to find the right version number on the changelog page, or the right function name of a page that documents several functions.

![Screenshot from 2023-07-21 08-41-54](https://github.com/TriplyDB/Documentation/assets/1685012/083a3d6b-49f7-4770-8139-92255338c2ee)

# Nice to have

- [x] Red is not a good color for in-line code. It looks too much like something is wrong ("error"), especially when scrolling through a page.

![Screenshot from 2023-07-21 08-32-03](https://github.com/TriplyDB/Documentation/assets/1685012/ed7c2889-d80e-4731-abff-a21289b2d8fe)

- [x] The vertical space between "Bugfixes:" and the item list of bugfixes is too big. It is currently equivalent to the vertical space between two block elements that are distinct (e.g. two pragraphs). But in an item lists, the introduction and the items belong together. This may not be fixable in ReadTheDocs, because it requires me to write an item list as follows:

```md
Bugfixes:

- Installation on Windows.
- Fixed warnings for a dependency (async-saxophone for XML processing), which did not yet officially support Node.js 18.
```

Instead of:

```md
Bugfixes:

- Installation on Windows.
- Fixed warnings for a dependency (async-saxophone for XML processing), which did not yet officially support Node.js 18.
```

So there is not way to distinguish a regular paragraph from a paragraph that introduces an item list.

![Screenshot from 2023-07-21 08-34-40](https://github.com/TriplyDB/Documentation/assets/1685012/e260d8e8-8222-4059-8fcb-19756119b97a)

This issue also exists in the real ReadThe Docs (e.g. <https://docs.conan.io/2/tutorial/consuming_packages/cross_building_with_conan.html>).

**Comment by wouterbeek** [comment]
- [x] Steps of instructions are not rendered.

In our current docs we use ordered lists with one or more blocks (paragraphs and code blocks) to guide a user through multiple steps: first import the library, then call this function, etc.

Mkdocs does not support code blocks in lists, and also does not number the list items properly:

![Screenshot from 2023-10-02 06-37-47](https://github.com/TriplyDB/Documentation/assets/1685012/e0e3b505-a001-4265-a1b1-45cbeaa261d3)

**Comment by wouterbeek** [comment]
- [x] No colorization for SPARQL or Turtle code blocks.

**Comment by wouterbeek** [comment]
- [x] The page top does not show where you are.

In the following screenshot, we are on page "root" > "TriplyETL" > "Assert" > "RATT" > "Statement". The breadcrumb only shows the highest level, with the home icon that links to "root", and the lowest level, i.e. "Statement".

![Screenshot from 2023-10-02 06-47-44](https://github.com/TriplyDB/Documentation/assets/1685012/9b0e1fcc-b554-4f58-800b-a0e20b453edf)

Expected: A better breadcrumb, showing where we are in the ToC tree structure.

Workaround: Put a head at the top of the page that includes terms from the invisible path (e.g. "RATT Statement Assertions" covers 3 steps in the current path).

**Comment by wouterbeek** [comment]
- [x] No in-page ToC.

It is difficult to find things within a page, since there is no in-page ToC. (Notice that there is a lot of empty screen real estate, ~45% on a laptop, on the right-hand side. This could be used to improve in-page navigation.)

**Comment by wouterbeek** [comment]
- [x] Difficult to see where a section/subsection/subsubsection starts.

The current subtle differences in header size are insufficient to comfortably communicate where a section starts, what is a subsection of a section, and what is a subsubsection of a subsection.

In the following screenshot "Dataset" is H1, this is a big section (may become its own page), but it is visually similar to the subsections (H2) underneath it ("Dataset.addPrefix") or the subsections under the previous section ("Asset.toStream"). Subsubsections (H3) are still very large.

Because of this, the page feels more like a list of items, not a tree structure. In print media there used to be numeric indicators, like "2. Dataset", "2.1 Dataset.addPrefix", and "2.1.1 Example".

![Screenshot from 2023-10-02 06-59-00](https://github.com/TriplyDB/Documentation/assets/1685012/4a549f63-dd4a-4f1d-ad4c-e9f310094a76)

**Comment by minatriply** [comment]
- [x] Sometimes, links may appear as regular text without the characteristic blue color or highlighting:

![i1 2](https://github.com/TriplyDB/Documentation/assets/106817499/69dc6e5f-b902-4db5-a761-29319c4ea1ec)

**Comment by minatriply** [comment]
- [x] The table should be of different scale.

![differentScaleTable](https://github.com/TriplyDB/Documentation/assets/106817499/88073a6e-6a38-4bbc-870b-adfbc58e3077)

**Comment by minatriply** [comment]
- [x] Mermaid Diagram Exceeds Documentation Boundaries
![tooBig](https://github.com/TriplyDB/Documentation/assets/106817499/e86b7336-1bb0-488f-8e85-9cb23474f13e)

**Comment by minatriply** [comment]
- [x]  When a sentence under the code starts with a dash, it appears very close to the code without any spacing.
- One solution might be to avoid using sentences with dashes, but is it possible to retain them and address the issue of sentences appearing immediately adjacent to the code?
- Example is from triplydb-js/app page

![errorWithCode](https://github.com/TriplyDB/Documentation/assets/106817499/cb665e02-9d3c-489b-9667-a5c0ca1ff14e)

**Comment by wouterbeek** [comment]
- [x] Heading 3 is currently in italic; this should be regular.
- [x] Heading 4 currently appears equally/comparably as big as heading 3; this should be clearly smaller.

**Comment by wouterbeek** [comment]
@mightymax at super speed!

**Comment by mightymax** [comment]
@wouter I have added numbering to headers, let me know if thus makes the structure of a page more clear. Note that pages must follow the correct headings without skipping  (also without numbering this is better to "understand" the document):
h1 > h2 > h3 > h4 > h5
[example page without h1](https://triplydb.github.io/Documentation/triply-api/)

**Comment by mightymax** [comment]
> * [x]  No in-page ToC.
> 
> It is difficult to find things within a page, since there is no in-page ToC. (Notice that there is a lot of empty screen real estate, ~45% on a laptop, on the right-hand side. This could be used to improve in-page navigation.)

It is now possible to render a TOC. To have a TOC, insert the text "[TOC]" at the top of your article. The TOC will only show on screens > 1100px wide.

[see this example](https://triplydb.github.io/Documentation/triply-etl/assert/ratt/statement/).

**Comment by wouterbeek** [comment]
> Note that pages must follow the correct headings without skipping

Wow, I have never seen any web-based framework follow header nesting in a logical way. If this is really how it works then that is of course great (but also very uncommon, so we need to actively tell authors to un-learn how all other websites work).

@mightymax Should we use `<h1>` for the title (i.e. once per page), or for the top-level headers (i..e once or more per page)? The title should be `<title>` in HTML5 (I think), but no website in the world works that way I think?

PS: The title is already given in metadata (at the top of the page), but websites do not use metadata to set the title I think.

**Comment by mightymax** [comment]
Every Dutch (and US, EU pending) public sector is required by [law](https://wetten.overheid.nl/BWBR0040936/2018-07-01) to follow correct headings. It is part of the [Web Content Accessibility Guidelines (WCAG)](https://www.w3.org/TR/WCAG21/) and [W3C Web Accessibility Initiative (WAI)](https://www.w3.org/WAI/). The WAI has a [document](https://www.w3.org/WAI/tutorials/page-structure/headings/) about best practices for headers. 

We should `<h1>` only at the top of a page and only once (see also [this recommendation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#avoid_using_multiple_h1_elements_on_one_page)).

For adding the title in metadata in `mkdocs`: there are two ways to do this: [YAML style](https://www.mkdocs.org/user-guide/writing-your-docs/#yaml-style-meta-data) and [MultiMarkdown Style](https://www.mkdocs.org/user-guide/writing-your-docs/#multimarkdown-style-meta-data)). But if you consider [these rules](https://www.mkdocs.org/user-guide/writing-your-docs/#multimarkdown-style-meta-data) you will see that the title in `nav` has the highest priority and thus metadata tags for `<title>` are ignored as are the `<h1>` headers. I have made a change in the template: the metadata page "title" will now overrule all other ways to determine a title. Note that providing this metadata is only needed should you wish to override the title provided in the navigation.

By the way: every title except on th home page will suffixed by " - Triply Documentation".

**Comment by wouterbeek** [comment]
@mightymax Nice that we can actually use HTML 5 tag in accordance with their meaning! I have been told for ~20 years that this is not possible, so I had given up on this. But good to hear that we do things correctly in Triply!

**Comment by wouterbeek** [comment]
- [ ] This one was reported (and fixed/checkmarked) earlier: I think that the vertical space between a paragraph and a list is too large:

![Screenshot from 2023-10-20 15-43-24](https://github.com/TriplyDB/Documentation/assets/1685012/a4e0630a-6901-4ed6-b6ca-95f8c234c517)

I'm not sure whether this is technologically possible to specify in CSS, but for me `<p>/<ul>` belongs together in a stronger way than `<p>/<p>` (where the vertical distance is legitimate).

**Comment by minatriply** [comment]
- [x] In the "[Generics](https://triplydb.github.io/Documentation/triply-etl/)" page of TriplyETL documentation, the mermaid diagram displays the "Enrich" label in an incorrect position


![Screenshot from 2023-10-26 10-32-14](https://github.com/TriplyDB/Documentation/assets/106817499/37470d39-ec17-46cf-aee3-930b4ff2fa7d)

**Comment by wouterbeek** [comment]
Let's close this one. The only open issue is too much space between a paragraph and a subsequent list, but it is acceptable.
